pkg/topology/query.go

```diff
@@ -156,100 +105,13 @@ func (p TopologyParams) getID() string {
-			(CASE WHEN (path IS NULL OR path = '') THEN id :: text ELSE concat(path,'.', id) END)
-			FROM components where id = :id)))`
-	} else {
-		s += ` and (parent.parent_id is null or starts_with(component_relationships.relationship_path, (SELECT
-			(CASE WHEN (path IS NULL OR path = '') THEN id :: text ELSE concat(path,'.', id) END)
-			FROM components where id = parent.id)))`
-	}
-	return s
-}
-
 func Query(params TopologyParams) (pkg.Components, error) {
-	query := fmt.Sprintf(`
-	SELECT json_agg(
-        jsonb_set_lax(
-            jsonb_set_lax(
-                jsonb_set_lax(
-                    jsonb_set_lax(
-                        jsonb_set_lax(
-                            jsonb_set_lax(
-                                to_jsonb(components),'{checks}', %s
-                            ), '{configs}', %s
-                        ), '{incidents}', %s
-                    ), '{insights}', %s
-                ), '{summary,insights}', %s
-            ), '{summary,incidents}', %s
-        )
-    ) :: jsonb AS components FROM components %s
-	UNION (
-    SELECT json_agg(
-        jsonb_set_lax(
-            jsonb_set_lax(
-                jsonb_set_lax(
-                    jsonb_set_lax(
-                        jsonb_set_lax(
-                            jsonb_set_lax(
-                                jsonb_set_lax(
-                                    to_jsonb(components), '{parent_id}', to_jsonb(component_relationships.relationship_id), true
-                                ),'{checks}', %s
-                            ), '{configs}', %s
-                        ), '{incidents}', %s
-                    ), '{insights}', %s
-                ), '{summary,insights}', %s
-            ), '{summary,incidents}', %s
-        )
-    ):: jsonb AS components FROM component_relationships
-    INNER JOIN components ON components.id = component_relationships.component_id
-    INNER JOIN components AS parent ON component_relationships.relationship_id = parent.id %s)`,
-		getChecksForComponents(), params.getConfigForComponents(), params.getIncidentsForComponents(), params.getConfigAnalysisForComponents(), params.getConfigAnalysisSummaryForComponents(), getIncidentSummaryForComponents(), params.GetComponentWhereClause(),
-		getChecksForComponents(), params.getConfigForComponents(), params.getIncidentsForComponents(), params.getConfigAnalysisForComponents(), params.getConfigAnalysisSummaryForComponents(), getIncidentSummaryForComponents(), params.GetComponentRelationWhereClause())
-
-	args := make(map[string]interface{})
-	if params.getID() != "" {
-		args["id"] = params.getID()
-	}
-	if params.Owner != "" {
-		args["owner"] = params.Owner
-	}
-	if params.Labels != "" {
-		args["labels"] = params.getLabels()
-	}
-
-	logger.Tracef("Querying topology (%s) => %s", params, query)
+	query, args := duty.TopologyQuery(duty.TopologyOptions{
```

Comment:
I think we should move all the filtering and unmarshalling into duty as well,

Reply:
I don't think we can do it right now without creating a cyclic dependency.

For duty to Marshal, CreateTreeStructure and Flatten it would need to have the components type which it will have to import from canary-checker.

We should eventually move it to duty but I would want to do that separately since the scope of that would be huge

Reply:
Can we not move the components type to duty, and then embed it into the canary-checker type ?

Reply:
Have created an issue to track this as discussed: https://github.com/flanksource/canary-checker/issues/793